dashboard/src/actions/metadataTreeActions.js

```diff
@@ -0,0 +1,104 @@
+import * as TYPES from "./types";
+
+import store from "store/store";
+
+const { getState } = store;
+
+// Helper functions
+const isChecked = (dataItem, checkedItems) =>
+  checkedItems && checkedItems.some((item) => item === dataItem.key);
+const areAllDescendantsChecked = (dataItem, checkedItems) =>
+  dataItem.children
+    ? dataItem.children.every((child) =>
+        areAllDescendantsChecked(child, checkedItems)
+      )
+    : isChecked(dataItem, checkedItems);
+const areSomeDescendantsChecked = (dataItem, checkedItems) =>
+  dataItem.children
+    ? dataItem.children.some((child) =>
+        areSomeDescendantsChecked(child, checkedItems)
+      )
+    : isChecked(dataItem, checkedItems);
+const setTreeViewChecked = (childNodes, isChecked) => {
+  childNodes.forEach(function iter(a) {
+    a.checkProps.checked = isChecked;
+    Array.isArray(a.children) && a.children.forEach(iter);
+  });
+};
+const getCheckedItemsKey = (ary) =>
+  ary.reduce(
+    (a, b) =>
+      a.concat(b.key, "children" in b ? getCheckedItemsKey(b.children) : []),
+    []
+  );
+const updateChildKeysList = (checked, checkedItems, childKeys) =>
+  checked
+    ? [...checkedItems, ...childKeys]
+    : checkedItems.filter((x) => !childKeys.includes(x));
+
+export const mapTree = (item) => {
+  const checkedItems = getState().overview.checkedItems;
+
+  item.checkProps.checked = areAllDescendantsChecked(item, checkedItems);
+  if (
+    !item.checkProps.checked &&
+    areSomeDescendantsChecked(item, checkedItems)
+  ) {
+    item.checkProps.checked = null;
+  }
+
+  const retVal = { ...item };
+  if (item.children) {
+    retVal.children = item.children.map((child) => mapTree(child));
+  }
+  return retVal;
+};
```

Comment:
What (I think) this code is doing is slightly odd:  for the current `item`, it decides whether it should be checked, unchecked, or "half-checked" -- based on the state of its children -- and then it proceeds to do the same recursively for each of the children...which means that it's going to visit each child once for each of its ancestors, which, if the tree is large, will be a lot of times!

What (I think) we should be doing is visiting the leaf nodes first, then determining from them how their parents should be set, and then returning to figure out how _their_ parents should be set, and so on.

Perhaps something like this would do it:
```javascript
export const mapTree = (item) => {
  const retVal = { ...item };
  if (item.children) {
    retVal.children = item.children.map((child) => mapTree(child));
    item.checkProps.checked = null;
    let seen = undefined;
    for (const c of item.children) {
      if (c.checkProps.checked == null) {
        return retVal;
      } else if (seen == undefined) {
        seen = c.checkProps.checked;
      } else if (seen !== c.checkProps.checked)
        return retVal;
      }
    }
    item.checkProps.checked = seen;
  } else {
    const checkedItems = getState().overview.checkedItems;
    item.checkProps.checked = isChecked(item, checkedItems);
  }

  return retVal;
};
```
That is, if the current `item` has children, evaluate each of the children (recursively).  Then, after setting `item`'s `checked` to `null`, look at the results of those evaluations:  if we find one which is `null`, then return immediately (with `item`'s set to `null`); otherwise, if we're looking at the first child (because `seen` is still `undefined`), remember whether it is set or not and move on to the next child; otherwise, if this child's `checked` is different from the last, then return immediately (with `item`'s set to `null`); otherwise, move on to the next child.  If we exit the loop without returning, then all the children are either checked or unchecked, and we set `item` to match.  Now, if `item` has no children, then we use `isChecked()` to set `checked`.  And, then we return.

So, we only call `mapTree()` once for each node, and we just look at the results on the way back up.  (We could replace my `for...of` loop with a `.every()` and a `.some()`, like you had, but that means going all the way through the children list twice, which is kind of unsatisfying -- with the `for...of` loop, we only check as many as we need to in order to get the right answer; and, this way, we don't need `areAllDescendantsChecked()` and `areSomeDescendantsChecked()` anymore.)

Reply:
That's really cool. Updated the code

Reply:
Glad you like it!  😁